data/trie/trieStorageManager.go

```diff
@@ -0,0 +1,326 @@
+package trie
+
+import (
+	"os"
+	"path"
+	"strconv"
+	"sync"
+
+	"github.com/ElrondNetwork/elrond-go/config"
+	"github.com/ElrondNetwork/elrond-go/core/check"
+	"github.com/ElrondNetwork/elrond-go/data"
+	"github.com/ElrondNetwork/elrond-go/data/mock"
+	"github.com/ElrondNetwork/elrond-go/hashing"
+	"github.com/ElrondNetwork/elrond-go/marshal"
+	"github.com/ElrondNetwork/elrond-go/storage"
+	"github.com/ElrondNetwork/elrond-go/storage/storageUnit"
+)
+
+// trieStorageManager manages all the storage operations of the trie (commit, snapshot, checkpoint, pruning)
+type trieStorageManager struct {
+	db            data.DBWriteCacher
+	pruningBuffer [][]byte
+
+	snapshots       []storage.Persister
+	snapshotId      int
+	snapshotDbCfg   config.DBConfig
+	snapshotsBuffer snapshotsBuffer
+
+	dbEvictionWaitingList data.DBRemoveCacher
+	storageOperationMutex sync.RWMutex
+}
+
+// NewTrieStorageManager creates a new instance of trieStorageManager
+func NewTrieStorageManager(db data.DBWriteCacher, snapshotDbCfg config.DBConfig, ewl data.DBRemoveCacher) (*trieStorageManager, error) {
+	if check.IfNil(db) {
+		return nil, ErrNilDatabase
+	}
+	if check.IfNil(ewl) {
+		return nil, ErrNilEvictionWaitingList
+	}
+
+	return &trieStorageManager{
+		db:                    db,
+		pruningBuffer:         make([][]byte, 0),
+		snapshots:             make([]storage.Persister, 0),
+		snapshotId:            0,
+		snapshotDbCfg:         snapshotDbCfg,
+		snapshotsBuffer:       newSnapshotsQueue(),
+		dbEvictionWaitingList: ewl,
+	}, nil
+}
+
+// Database returns the main database
+func (tsm *trieStorageManager) Database() data.DBWriteCacher {
+	tsm.storageOperationMutex.Lock()
+	defer tsm.storageOperationMutex.Unlock()
+
+	return tsm.db
+}
+
+// SetDatabase sets the provided database as the main database
+func (tsm *trieStorageManager) SetDatabase(db data.DBWriteCacher) {
+	tsm.storageOperationMutex.Lock()
+	tsm.db = db
+	tsm.storageOperationMutex.Unlock()
+}
+
+// Clone returns a new instance of trieStorageManager
+func (tsm *trieStorageManager) Clone() data.StorageManager {
+	tsm.storageOperationMutex.Lock()
```

Comment:
I think you can replace with something like:
``` GOLANG
tsmCopy := *tsm
tsm.storageOperationMutex.Lock()
tsmCopy.snasphotsBuffer = tsm.snapshotsBuffer.clone()
tsm.storageOperationMutex.Unlock()

return &tsmCopy
```

check if it will work this way

Reply:
I think the way it is right now is a bit cleaner.